File: app/src/main/java/com/barryyang/barryyangdemo/CircularRevealHelper.java
package com.barryyang.barryyangdemo;

import android.animation.Animator;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewAnimationUtils;

import androidx.constraintlayout.widget.ConstraintHelper;
import androidx.constraintlayout.widget.ConstraintLayout;

public class CircularRevealHelper extends ConstraintHelper {

    public CircularRevealHelper(Context context) {
        super(context);
    }

    public CircularRevealHelper(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CircularRevealHelper(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public void updatePostLayout(ConstraintLayout container) {
        super.updatePostLayout(container);
        View[] views = getViews(container);
        for (View v : views) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                double hypot = Math.hypot(v.getHeight() / 2.0d, v.getWidth() / 2.0d);
                Animator circularReveal = ViewAnimationUtils.createCircularReveal(v, v.getWidth() / 2, v.getHeight() / 2, 0, (float) hypot);
                circularReveal.setDuration(3000);
                circularReveal.start();
            }
        }
    }
}

File: app/src/main/java/com/barryyang/barryyangdemo/android/constraint/CircularRevealHelper.kt
package com.barryyang.barryyangdemo.android.constraint

import android.animation.Keyframe
import android.animation.ObjectAnimator
import android.animation.PropertyValuesHolder
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.view.animation.AnticipateOvershootInterpolator
import androidx.constraintlayout.widget.ConstraintHelper
import androidx.constraintlayout.widget.ConstraintLayout

class CircularRevealHelper : ConstraintHelper {

    constructor(context: Context?) : super(context) {}
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    override fun updatePostLayout(container: ConstraintLayout) {
        super.updatePostLayout(container)
        val views = getViews(container)
        views.forEach {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                startEnterAnimation(it)
            }
        }
    }

    private fun startEnterAnimation(view: View) {
        val translationY = -100f
        view.translationY = translationY
        val translationYHolder = PropertyValuesHolder.ofFloat(View.TRANSLATION_Y, translationY, 0f)
        val keyFrame1 = Keyframe.ofFloat(0f, -6f)
        val keyFrame2 = Keyframe.ofFloat(0.6f, 25f)
        val keyFrame3 = Keyframe.ofFloat(1f, 0f)
        val rotateHolder = PropertyValuesHolder.ofKeyframe(View.ROTATION, keyFrame1, keyFrame2, keyFrame3)
        ObjectAnimator.ofPropertyValuesHolder(view, translationYHolder, rotateHolder)
                .apply { interpolator = AnticipateOvershootInterpolator() }
                .setDuration(600L)
                .start()
    }
}